### PR TITLE
Add support for unix domain sockets

### PR DESCRIPTION
**Pull request by bwoebi**
Including:
- default unix socket for dogstatsd
- default unix socket for background sender

Manually tested dogstatsd by creating a dummy socket `socat UNIX-LISTEN:/var/run/datadog/dsd.socket -` - not sure how to test that automatically, or against the real stats receiver.

### Readiness checklist
- [x] (only for Members) Changelog has been added to the release document.
- [x] Tests added for this feature/bug.

### Reviewer checklist
- [ ] Appropriate labels assigned.
- [ ] Milestone is set.
- [ ] Changelog has been added to the release document. For community contributors the reviewer is in charge of this task.

### CODE REVIEW

File: tests/ext/background-sender/agent_headers_unix_domain_socket.phpt [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+DD_TRACE_AGENT_FLUSH_AFTER_N_REQUESTS=1
+DD_TRACE_AGENT_FLUSH_INTERVAL=333
+DD_TRACE_GENERATE_ROOT_SPAN=0
+--FILE--
+<?php
+include __DIR__ . '/../includes/request_replayer.inc';
+
+$socketPath = getenv("DD_TRACE_AGENT_URL");
+@unlink($socketPath);
+$p = popen(PHP_BINARY . ' -r \'$s=stream_socket_server("' . $socketPath . '");print 1;$c=stream_socket_accept($s);$t=stream_socket_client("request-replayer:80");$a=$r=[$c,$t];foreach($r as $f)stream_set_blocking($f, false);while(stream_select($r,$w,$e,null)){$d=fread($f=reset($r),4096);if($d=="")exit;fwrite($f==$t?$c:$t,$d);$r=$a;}\'', 'r');
```
Comment (morrisonlevi) [inline tests/ext/background-sender/agent_headers_unix_domain_socket.phpt:17]: Must the `-r` value contain no newlines?

File: ext/php8/coms.c [hunk 2]
```diff
@@ -720,15 +721,30 @@ char *ddtrace_agent_url(void) {
...
 }
 
 void ddtrace_curl_set_hostname(CURL *curl) {
     char *url = ddtrace_agent_url();
     if (url && url[0]) {
-        size_t agent_url_len = strlen(url) + sizeof(TRACE_PATH_STR);
+        char *http_url = url;
+        if (strlen(url) > 7 && strncmp(url, "unix://", 7) == 0) {
+            curl_easy_setopt(curl, CURLOPT_UNIX_SOCKET_PATH, url + 7);
+            http_url = "http://localhost";
```
Comment (morrisonlevi) [inline ext/php8/coms.c:743]: For curl, when using UDS, does it matter what the host is? Could `localhost` be replaced with theoretically anything?

File: tests/ext/background-sender/agent_headers_unix_domain_socket.phpt (hunk 1 above)
Comment (bwoebi) [reply tests/ext/background-sender/agent_headers_unix_domain_socket.phpt:17]: Can you please suggest a proper formatting? It's just a line to forward proxy data from one socket to the other.

Comment (morrisonlevi) [reply tests/ext/background-sender/agent_headers_unix_domain_socket.phpt:17]: I would probably use a heredoc and format it basically as normal PHP, I think? I get that it is "unimportant" but I'd still like to read and understand it :)

Comment (morrisonlevi) [reply tests/ext/background-sender/agent_headers_unix_domain_socket.phpt:17]: Also, it appears to be used more than once, so maybe make a function for it?

File: ext/php5/dogstatsd_client.c [hunk 3]
```diff
@@ -9,38 +9,47 @@
 ZEND_EXTERN_MODULE_GLOBALS(ddtrace);
 
 #define METRICS_CONST_TAGS "lang:php,lang_version:" PHP_VERSION ",tracer_version:" PHP_DDTRACE_VERSION
+#define DEFAULT_UDS_PATH "/var/run/datadog/dsd.socket"
```
Comment (r1viollet) [inline ext/php5/dogstatsd_client.c:12]: the usual env var available on the machine (connecting to a random pod) is the following 
```
STATSD_URL=unix:///var/run/datadog-agent/statsd.sock
```

**Comment by r1viollet** [comment]
Testing with a real Datadog agent could be a good way to test this.

File: ext/php5/dogstatsd_client.c (hunk 3 above)
Comment (bwoebi) [reply ext/php5/dogstatsd_client.c:12]: I haven't added configurability of the dogstatsd unix socket here. (I.e. we don't support STATSD_URL here at all.)
It just defaults to the unix socket if there's one.

Unlike with the tracer (DD_TRACE_AGENT_URL), I wasn't aware of a "full" url env var for dogstatsd. So `STATSD_URL` is the proper env var for that?

Comment (r1viollet) [reply ext/php5/dogstatsd_client.c:12]: I'm honestly confused by the RFC. This was the only statsd socket I was aware of on a POD. I'll ask around and see if I can figure this one out.

Comment (r1viollet) [reply ext/php5/dogstatsd_client.c:12]: The internal statsd configuration is different from the one defined in public documentations.

Comment (bwoebi) [reply ext/php5/dogstatsd_client.c:12]: It's now configurable via DD_AGENT_HOST, so can also be used on other non-standard paths.

File: src/dogstatsd/client.c [hunk 4]
```diff
@@ -60,14 +61,25 @@ dogstatsd_client dogstatsd_client_ctor(struct addrinfo *addrs, char *buffer,
     }
   }
 
+  if (addr->ai_family == PF_UNIX) {
+    if (!connect(client.socket, addr->ai_addr, addr->ai_addrlen)) {
```
[the remark this answers is not in the document]
Comment (bwoebi) [reply src/dogstatsd/client.c:65]: That case does not matter, it then does connect(-1) which just fails again here.

**Review by morrisonlevi — APPROVED** [review]
The code looks good to me, and I was able to send traces using the real agent. I committed the necessary changes to `docker-compose.yml` so this will work out-of-the-box.

**Pull request merged**